hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/ClusterJoinManager.java

```diff
@@ -794,11 +800,12 @@ private void startJoin(OnJoinOp preJoinOperation) {
                     return;
                 }
 
-                OnJoinOp preJoinOp = preparePreJoinOps();
+                // Run all joining members' provided pre join operations now, but only
+                //  execute them locally (do not broadcast to other members)
+                runProvidedPostJoinOpsWithoutBroadcast();
 
-                if (preJoinOperation != null) {
-                    nodeEngine.getOperationService().run(preJoinOperation);
-                }
+                // Prepare our normal pre-join operations, which will be broadcast remotely
+                OnJoinOp preJoinOp = preparePreJoinOps();
```

Comment:
I think it's worth adding a comment here that pre-join op from `master` to joining members must be prepared *after* pre-join ops from all the joining members in the batch are applied to `master` member.

Reply:
Good call, updated the comment here: https://github.com/hazelcast/hazelcast/pull/25114/commits/2111c7133c536d66d50c14e750e15e4cbf01c0d7 👍